File: destack-language/destack/core/space/folder.py
from typing import TYPE_CHECKING

from ..builtin import (
    Entity,
    EnumType,
    NodeType,
    OptionEnum,
    TraitType,
    declare_entity,
    declare_enum,
    declare_option,
    declare_property,
)

if TYPE_CHECKING:
    pass


@declare_enum(EnumType.FOLDER_TYPE)
class FolderType(OptionEnum):
    SYSTEM = declare_option(1, "Root", description="The root folder of a Space")
    HOME = declare_option(2, "Home", description="The home folder of a Space")
    GENERAL = declare_option(3, "General", description="A general folder")
    MODULE = declare_option(4, "Module", description="A module")
    APP = declare_option(5, "App", description="An app folder")
    # SERVICE, PLUGIN, WIDGET, TEMPLATE, LIBRARY, ...


@declare_entity(
    NodeType.FOLDER,
    traits=(
        TraitType.OWNABLE,
        TraitType.ORDERED,
        TraitType.JOINABLE,
        TraitType.STARABLE,
        TraitType.FOLLOWABLE,
        TraitType.REACTABLE,
    ),
)
class Folder(Entity):
    """A Folder is a sub-space of a Space."""

    type: FolderType = declare_property(
        100,
        is_repr=True,
        default=FolderType.GENERAL,
        tag=None,
    )
    slug: str | None = declare_property(103, is_repr=True, tag=None)

File: destack-rs/destack/src/core/space/folder.rs
//! destack.core.space.folder@2025.08.15.1

#![destack::partial(destack.core.space.folder, file)]

#[destack::generated(FolderType, enum, block)]
/// FolderType
pub enum FolderType {
    /// The root folder of a Space
    System = 1,
    /// The home folder of a Space
    Home = 2,
    /// A general folder
    General = 3,
    /// A module
    Module = 4,
    /// An app folder
    App = 5,
}
